date_range/models/date_range_type.py

```diff
@@ -22,7 +24,24 @@ def _default_company(self):
     company_id = fields.Many2one(
         comodel_name='res.company', string='Company', index=1,
         default=_default_company)
+    parent_type_id = fields.Many2one(
+        comodel_name='date.range.type',
+        index=1)
 
     _sql_constraints = [
         ('date_range_type_uniq', 'unique (name,company_id)',
          'A date range type must be unique per company !')]
+
+    @api.constrains('parent_type_id')
+    def _validate_parent_type_id(self):
+        for record in self:
+            parent = record
+            while parent:
+                if not parent.parent_type_id:
+                    break
+                if record.parent_type_id == parent:
+                    raise ValidationError(
+                        _("A type parent  can not have a parent:"
```

Comment:
The message is wrong. A type with a parent can have a parent. Like months can belong to years and years can belong to centuries. So the error should be something like: type can not have itself as parent or ancestor. No need to have parent.name and record.name both in the message. The moment the error occurs, they will be the same! Actually no need to have variables at all in the message, the moment the error occurs the user is on the record, so needs no help in identifying the record.